### PR TITLE
RATEST-294: User Role Based Login

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-294

### CODE REVIEW

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature [hunk 1]
```diff
@@ -13,5 +13,4 @@ Feature: Reports Management
     And user clicks on run button from manage Reports page
     And user enters start date
     And user enters end date
-    And user clicks on request report button
-    Then user clicks on view report link
+    # Then user clicks on view report link
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:16]: any reason for commenting out this step?

**Comment by sherrif10** [comment]
cc @kdaud  Looking forward for reviews.

File: qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties [hunk 2]
```diff
@@ -1,9 +1,8 @@
-webapp.url=http://localhost:8080/openmrs
+webapp.url=https://qa-refapp.openmrs.org/openmrs
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties:1]: This change was because of updating the branch to the current master branch.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 3]
```diff
@@ -0,0 +1,10 @@
+Feature: User based login
+
+@selenium
+@defaultLogin
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:4]: how about renaming the tag  to `@login` ?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 4]
```diff
@@ -0,0 +1,10 @@
+Feature: User based login
+
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:2]: would be a good implementation if we have a background of the scenario.

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:2]: For some reasons. We dont need a back ground scenario here. One reason is because method called in before method does it for all the initiating the page. in this case what would go in background case would ideally be intiating home page and stuff which is by default done by initiateLogin. thats the reason. Let me know if you need more clarification.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 3 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:4]: Login is being used some where here  https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java#L27   so they will contradict. cucumber allows one single naming.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
+package org.openmrs.contrib.qaframework.automation;
+
+import io.cucumber.java.After;
+import io.cucumber.java.Before;
+import io.cucumber.java.en.And;
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
+
+import static org.junit.Assert.assertTrue;
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:9]: checkout the existing import order and align

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java]: Resolved thnks

**Comment by sherrif10** [comment]
Resolved thanks.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java [hunk 6]
```diff
@@ -39,43 +43,35 @@ public void verifyModulesAvailableOnHomePage() {
...
-    @Test
-    @Category(BuildTests.class)
+    @Then("user verifies modules available on home page after login as sysadmin")
     public void verifySysadminModulesAvailableOnHomePage() {
         goToLoginPage().loginAsSysadmin();
-        initiateHomePage();
         assertTrue(homePage.isAppointmentSchedulingAppPresent());
         assertTrue(homePage.isSystemAdministrationAppPresent());
-    }
+    }
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/VerifyLoginSteps.java:76]: you can remove the introduced white spaces.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature [hunk 7]
```diff
@@ -14,4 +14,4 @@ Feature: Reports Management
     And user enters start date
     And user enters end date
     And user clicks on request report button
-    Then user clicks on view report link
+    Then user clicks on view report link
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:17]: is this change needed?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:17]: No changes introduced here. its been resorted back on the same line.

**Comment by sherrif10** [comment]
Hello @kdaud  is this worth approving and we merge it.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature [hunk 8]
```diff
@@ -14,4 +14,4 @@ Feature: Reports Management
     And user enters start date
     And user enters end date
     And user clicks on request report button
-    #Then user clicks on view report link
+    #Then user clicks on view report link
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:17]: is this change part of the ticket requirement?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:17]: This change is part of the  requirement as its stated above.

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:17]: are the steps in report feature file used by Login steps definition?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:17]: Please comfirm with the changes on master. this as you see it as a change was because of updating my branch to master. checkout this line https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature#L17

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:17]: > this as you see it as a change was because of updating my branch to master

@sherrif10 how do you update your branch?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:17]: uhmm is this of help https://gist.github.com/whoisryosuke/36b3b41e738394170b9a7c230665e6b9

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:17]: When master changes are merged into a branch following the documentation you have shared, the changes are not reflected in the branch.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+@selenium
+@defaultLogin
+Scenario Outline: User Based login verification
+   When user verifies modules available on home page
+   And user verifies modules available on homepage after login as clerk
+   And user verifies modules available on home page after login as Docter
+   And user verifies modules available on home page after login as Nurse
+   And user verifies modules available on home page after login as sysadmin
+   Then system goes back to login page
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:11]: Add an empty line at the EOF to cater for git collaborators

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature]: Resolved.Thanks.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 10]
```diff
@@ -0,0 +1,12 @@
+Feature: User based login
+
+@selenium
+@defaultLogin
+Scenario Outline: User Based login verification
+   When user verifies modules available on home page
+   And user verifies modules available on homepage after login as clerk
+   And user verifies modules available on home page after login as Docter
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:8]: @sherrif10 what is thought about `Docter` vs `Doctor`?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+   And user verifies modules available on home page after login as Nurse
+   And user verifies modules available on home page after login as sysadmin
+   Then system goes back to login page
+
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:12]: If you are able to remove the white spaces in the last line will be sweet!

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature [hunk 12]
```diff
@@ -15,3 +15,4 @@ Feature: Reports Management
     And user enters end date
     And user clicks on request report button
     #Then user clicks on view report link
+
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature:18]: Any reason for adding white spaces in the last empty line?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 11 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature]: Resolved.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature (hunk 12 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/015-reports/reports.feature]: Resolved

File: qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties [hunk 13]
```diff
@@ -1,4 +1,4 @@
...
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties]: Ohh sorry. Let me correct that.

[the remark this answers is not in the document]
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties]: Resolved thanks man i dont how it got in that pr.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 9 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:11]: @kdaud  Let me know if this looks fine.

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:11]: @sherrif10 take a second look at the comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/342/files#r806469786)

File: qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties [hunk 14]
```diff
@@ -7,4 +7,4 @@ login.auto=false
 headless=true
 db.host=openmrs
 webdriver.gecko.driver=firefoxdriver/linux/geckodriver
-includes.csrftoken=true
+includes.csrftoken=true
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties:10]: any reason for this change?

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties:10]: Is this not part of the master.https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties#L10

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties:10]: any reason for removing the last empty line?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 15]
```diff
@@ -0,0 +1,11 @@
+Feature: User based login
+
+@selenium
+@defaultLogin
+Scenario Outline: User Based login verification
+   When user verifies modules available on home page
+   And user verifies modules available on homepage after login as clerk
+   And user verifies modules available on home page after login as Doctor
+   And user verifies modules available on home page after login as Nurse
+   And user verifies modules available on home page after login as sysadmin
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:10]: isn't it better to have the right work for `sysadmin`?

File: qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties (hunk 14 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties:10]: @kdaud  where is the last empty line removed here.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 15 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:10]: Would like to keep it the way it was initially.

File: qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties (hunk 14 above)
Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties:10]: checkout master on your local repo and see the file within your IDE!

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 15 above)
Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:10]: system adminstration .

**Comment by sherrif10** [comment]
Wondering  whether @kdaud  is watching this guy.

File: qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties (hunk 14 above)
Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties:10]: isn't there an empty line at the EOF in master?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature [hunk 16]
```diff
@@ -0,0 +1,11 @@
+Feature: User based login
+
+@selenium
+@defaultLogin
+Scenario Outline: User Based login verification
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:5]: when do we use `Scenario Outline` over `Scenario`?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature (hunk 9 above)
Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/02-login/verifyLogin.feature:11]: @sherrif10 did you see the comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/342/files#r809015231)
